Properly handle HTTP request failures instead of raising JSON decode errors

smurching opened the pull request:
### Current state of the world
We make all HTTP requests in the Python client via a `mlflow.utils.rest_utils.http_request` helper that returns a Response object unless the request fails with a server-level (5xx) error. Callers of `http_request` are responsible for checking whether the request failed e.g. with an application-level error. Currently, we don't consistently perform this check (e.g. in `DbfsArtifactRepository` and the `RestStore`), and instead attempt to parse the response body of a failed request, which results in somewhat confusing JSON decode errors. When we do perform the check, it's through error-handling logic spread around different modules.

### Updates in this PR
This PR consolidates exception-handling logic for HTTP requests into a single `http_request_safe` method, which wraps the old `http_request` helper but also verifies that the response status is 200. After this change, calling `log_artifact` or `log_param` against DBFS / a Databricks tracking server with bad auth will yield well-formed error messages instead of JsonDecodeErrors.

The goal of this PR is to make using `http_request_safe` (& by extension error-handling of failed requests) the default. A few methods (namely, listing files / getting the status of files on DBFS) still use `http_request` since non-200 responses are useful in indicating whether or nor files exist, but no other methods do.

smurching committed:
WIP

smurching committed:
Refactor common HTTP-request logic into new http_request_safe method, call it everywhere + add test

smurching committed:
Fix tests

smurching commented on mlflow/projects/databricks.py:
```diff
@@ -218,13 +208,11 @@ def get_run_result_state(self, databricks_run_id):
     def jobs_runs_cancel(self, databricks_run_id):
         response = self._databricks_api_request(
             endpoint="/api/2.0/jobs/runs/cancel", method="POST", json={"run_id": databricks_run_id})
-        _check_response_status_code(response)
```
No need to check this since we perform the necessary check in `http_request_safe` (called from self._databricks_api_request)

smurching committed:
Check response status where we still use http_request

smurching commented on mlflow/store/dbfs_artifact_repo.py:
```diff
@@ -28,12 +28,14 @@ def __init__(self, artifact_uri, get_host_creds):
     def _dbfs_list_api(self, json):
-        return self._databricks_api_request(endpoint=LIST_API_ENDPOINT, method='GET', json=json)
+        host_creds = self.get_host_creds()
+        return http_request(
```
As in `databricks.py`, here we still use the old `http_request` API since non-200 response statuses can indicate that no files exist under the specified DBFS path

smurching commented on mlflow/projects/databricks.py:
```diff
@@ -74,18 +66,24 @@ def _upload_to_dbfs(self, src_path, dbfs_fuse_uri):
-            endpoint="/api/2.0/dbfs/get-status", method="GET", json={"path": "/%s" % dbfs_path})
-        json_response_obj = json.loads(response.text)
+        host_creds = databricks_utils.get_databricks_host_creds(self.databricks_profile)
+        response = rest_utils.http_request(
```
Here we still use the old `http_request` API since non-200 response statuses can indicate that no file exists at the specified DBFS path

smurching committed:
Restore RestException

smurching committed:
Updates after rebase

smurching committed:
Make test more precise

andrewmchen commented on mlflow/utils/rest_utils.py:
```diff
@@ -13,6 +14,25 @@
 RESOURCE_DOES_NOT_EXIST = 'RESOURCE_DOES_NOT_EXIST'
 
 
+class RestException(MlflowException):
```
Hm, maybe a merge conflict? Can we use the `RestException` in `mlflow.exceptions`?

andrewmchen commented on mlflow/utils/rest_utils.py:
```diff
@@ -13,6 +14,25 @@
+        self.json = json
+
+
+def check_response_status(response):
```
What are the usages of this guy?

smurching committed:
Address some comments

andrewmchen commented on mlflow/store/dbfs_artifact_repo.py:
```diff
@@ -69,14 +71,8 @@ def log_artifact(self, local_file, artifact_path=None):
+            self._databricks_api_request(
                 endpoint=http_endpoint, method='POST', data=f, allow_redirects=False)
-            if response.status_code == 409:
-                raise MlflowException('File already exists at {} and can\'t be overwritten.'
```
Did we check what the error from the server is in this case?

smurching replied:
Good question, we end up logging it, though it's admittedly kind of ugly:

```
Traceback (most recent call last):
  File "/Users/siddharthmurching/code/mlflow/mlflow/utils/rest_utils.py", line 66, in http_request_safe
    js_dict = json.loads(response.text)
  File "/Users/siddharthmurching/miniconda3/envs/mlflow_python3/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/Users/siddharthmurching/miniconda3/envs/mlflow_python3/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Users/siddharthmurching/miniconda3/envs/mlflow_python3/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "examples/quickstart/mlflow_tracking.py", line 21, in <module>
    log_artifacts("outputs")
  File "/Users/siddharthmurching/code/mlflow/mlflow/tracking/fluent.py", line 223, in log_artifacts
    MlflowClient().log_artifacts(run_id, local_dir, artifact_path)
  File "/Users/siddharthmurching/code/mlflow/mlflow/tracking/client.py", line 166, in log_artifacts
    artifact_repo.log_artifacts(local_dir, artifact_path)
  File "/Users/siddharthmurching/code/mlflow/mlflow/store/dbfs_artifact_repo.py", line 91, in log_artifacts
    endpoint=endpoint, method='POST', data=f, allow_redirects=False)
  File "/Users/siddharthmurching/code/mlflow/mlflow/store/dbfs_artifact_repo.py", line 33, in _databricks_api_request
    return http_request_safe(host_creds=host_creds, endpoint=endpoint, **kwargs)
  File "/Users/siddharthmurching/code/mlflow/mlflow/utils/rest_utils.py", line 68, in http_request_safe
    raise MlflowException("%s. Response body: '%s'" % (base_msg, response.text))
mlflow.exceptions.MlflowException: API request to endpoint /dbfs/databricks/mlflow/0/038ce380607e4f26a6e86ccc9a5e84be/artifacts/test.txt failed with error code 409 != 200. Response body: '<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 409 </title>
</head>
<body>
<h2>HTTP ERROR: 409</h2>
<p>Problem accessing /dbfs/databricks/mlflow/0/038ce380607e4f26a6e86ccc9a5e84be/artifacts/test.txt. Reason:
<pre>    File already exists, cannot overwrite: &apos;/databricks/mlflow/0/038ce380607e4f26a6e86ccc9a5e84be/artifacts/test.txt&apos;</pre></p>
<hr />
</body>
</html>
'
```

smurching replied:
Edit: pushed an update that hides the JSONDecodeError part of the exception

```
Running mlflow_tracking.py
Traceback (most recent call last):
  File "examples/quickstart/mlflow_tracking.py", line 21, in <module>
    log_artifacts("outputs")
  File "/Users/siddharthmurching/code/mlflow/mlflow/tracking/fluent.py", line 223, in log_artifacts
    MlflowClient().log_artifacts(run_id, local_dir, artifact_path)
  File "/Users/siddharthmurching/code/mlflow/mlflow/tracking/client.py", line 166, in log_artifacts
    artifact_repo.log_artifacts(local_dir, artifact_path)
  File "/Users/siddharthmurching/code/mlflow/mlflow/store/dbfs_artifact_repo.py", line 91, in log_artifacts
    endpoint=endpoint, method='POST', data=f, allow_redirects=False)
  File "/Users/siddharthmurching/code/mlflow/mlflow/store/dbfs_artifact_repo.py", line 33, in _databricks_api_request
    return http_request_safe(host_creds=host_creds, endpoint=endpoint, **kwargs)
  File "/Users/siddharthmurching/code/mlflow/mlflow/utils/rest_utils.py", line 74, in http_request_safe
    raise MlflowException("%s. Response body: '%s'" % (base_msg, response.text))
mlflow.exceptions.MlflowException: API request to endpoint /dbfs/databricks/mlflow/0/ae3c099bc3cf44f696102d44c615fcc1/artifacts/test.txt failed with error code 409 != 200. Response body: '<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>
<title>Error 409 </title>
</head>
<body>
<h2>HTTP ERROR: 409</h2>
<p>Problem accessing /dbfs/databricks/mlflow/0/ae3c099bc3cf44f696102d44c615fcc1/artifacts/test.txt. Reason:
<pre>    File already exists, cannot overwrite: &apos;/databricks/mlflow/0/ae3c099bc3cf44f696102d44c615fcc1/artifacts/test.txt&apos;</pre></p>
<hr />
</body>
</html>
'
```

andrewmchen replied:
Yeah I agree with you here.

smurching committed:
Update exception handling to not raise an MLflowException while handling another exception
(if we do that, the stacktrace will still show the JsonDecodeError in Python 3)

smurching committed:
Fix lint

smurching commented:
Thanks for the review @andrewmchen! Merging :)
